compiler/fir/resolve/src/org/jetbrains/kotlin/fir/types/ConeTypeCompatibilityChecker.kt

```diff
@@ -0,0 +1,410 @@
+        computeIfAbsent(parameter) { BoundTypeArguments(mutableSetOf(), mutableSetOf()) }.let {
+            val type = boundTypeArgument.type
+            if (boundTypeArgument.variance.allowsInPosition) {
+                it.lower += type.collectLowerBounds()
+            }
+            if (boundTypeArgument.variance.allowsOutPosition) {
+                it.upper += type.collectUpperBounds()
+            }
+        }
+    }
+
+    private fun FirClassLikeDeclaration<*>.getSuperTypes(): List<ConeClassLikeType> {
+        return when (this) {
+            is FirTypeAlias -> listOfNotNull(expandedTypeRef.coneTypeSafe())
+            is FirClass<*> -> superTypeRefs.mapNotNull { it.coneTypeSafe() }
+            else -> emptyList()
+        }
+    }
+
+    private fun ConeClassLikeType.getClassLikeElement(ctx: ConeInferenceContext): FirClassLikeDeclaration<*>? =
+        ctx.symbolProvider.getSymbolByLookupTag(lookupTag)?.fir
+
+    private fun FirClassLikeDeclaration<*>.getTypeParameter(index: Int): FirTypeParameterRef? {
+        return when (this) {
+            is FirTypeAlias -> typeParameters[index]
+            is FirClass<*> -> typeParameters[index]
+            else -> return null
+        }
+    }
+
+    /** A class declaration and the arguments bound to the declared type parameters. */
+    private data class TypeArgumentMapping(
+        val typeParameterOwner: FirClassLikeDeclaration<*>,
+        val mapping: Map<FirTypeParameterRef, BoundTypeArgument>
+    )
+
+    /** A single bound type argument to a type parameter declared in a class. */
+    private data class BoundTypeArgument(val type: ConeKotlinType, val variance: Variance)
+
+    /** Accumulated type arguments bound to a type parameter declared in a class. */
+    private data class BoundTypeArguments(val upper: MutableSet<ConcreteType>, val lower: MutableSet<ConcreteType>)
+
+    private sealed class ConcreteType {
+        data class ClassLikeType(val type: ConeClassLikeType) : ConcreteType()
+        object IntegerLiteralType : ConcreteType()
+
+        fun toConcreteClass(ctx: ConeInferenceContext): ConcreteClass? {
+            return when (this) {
+                is ClassLikeType -> type.lookupTag.toConcreteClass(ctx)
+                IntegerLiteralType -> ConcreteClass.IntegerLiteral
+            }
+        }
+    }
+
+    private sealed class ConcreteClass {
+        data class Class(val firClass: FirClass<*>) : ConcreteClass()
+        data class TypeAlias(val firTypeAlias: FirTypeAlias, val referencedClass: FirClass<*>) : ConcreteClass()
+        object IntegerLiteral : ConcreteClass()
+
+        fun isInterface(): Boolean {
+            return when (this) {
+                is Class -> firClass.classKind == ClassKind.INTERFACE
+                is TypeAlias -> referencedClass.classKind == ClassKind.INTERFACE
+                IntegerLiteral -> false
+            }
+        }
+
+        fun getSuperConcreteClasses(ctx: ConeInferenceContext): List<ConcreteClass> {
```

Comment:
I think this and `getThisAndAllSuperConcreteClasses` need to be cached (via `lazy`) to avoid multiple traversals. To do so, we may need a canonical instance per `FirClass<*>` and/or `FirTypeAlias` via `ConcreteClassFactory` or `ConcreteClassProvider`.

Reply:
Sounds good! Will do.